File: CONCEPTS/class_example.R
library(S7)

df <- tibble::tribble(
  ~GAME_ID, ~TEAM_ID, ~PLAYER_ID, ~FGM, ~FG3M, ~FTM,
      1000,      100,          1,   10,    12,   12,
      1000,      100,          2,    4,     4,    7,
      1000,      200,          3,    2,     6,    5,
      1000,      200,          4,    8,     2,    7,
      2000,      100,          1,   10,     4,   10,
      2000,      100,          2,   11,     5,    4,
      2000,      300,          5,    8,    10,    9,
      2000,      300,          6,    7,     6,    3
)

#--------------------- Boxscore class ----------------------------------------
Boxscore <- new_class("Boxscore",
  properties = list(
    data = class_data.frame
  ),
  validator = function(self) {
    required_cols <- c("GAME_ID", "TEAM_ID", "PLAYER_ID", "FGM", "FG3M", "FTM")
    missing <- setdiff(required_cols, names(self@data))
    if (length(missing) > 0) {
      paste("Missing required columns:", paste(missing, collapse = ", "))
    }
  }
)

method(print, Boxscore) <- function(x, ...) {
  n_rows <- nrow(x@data)
  n_games <- length(unique(x@data$GAME_ID))
  cat(sprintf("Boxscore: %d rows, %d games\n\n", n_rows, n_games))
  print(x@data)
  invisible(x)
}

#--------------------- main program  ----------------------------------------
box <- Boxscore(data = df)
print(box)

File: CONCEPTS/class_example.py
import pandas as pd

# --------------------- Boxscore class ----------------------------------------
class Boxscore:
    """Represents game-level statistics."""
    def __init__(self, data: pd.DataFrame):
        self._data = data
        self._validate()

    def _validate(self):
        """Validates that all required columns are present."""
        required_cols = ["GAME_ID", "TEAM_ID", "PLAYER_ID", "FGM", "FG3M", "FTM"]
        missing = set(required_cols) - set(self._data.columns)
        if missing:
            raise ValueError(f"Missing required columns: {', '.join(missing)}")

    @property
    def data(self) -> pd.DataFrame:
        """Property to access the underlying data frame."""
        return self._data

    def __repr__(self):
        """Provides a custom string representation for the class."""
        n_rows = len(self.data)
        n_games = self.data['GAME_ID'].nunique()
        header = f"Boxscore: {n_rows} rows, {n_games} games\n\n"
        return header + self.data.to_string()

# --------------------- main program ----------------------------------------

# Create the initial DataFrame
df = pd.DataFrame({
    'GAME_ID':    [1000, 1000, 1000, 1000, 2000, 2000, 2000, 2000],
    'TEAM_ID':    [100,  100,  200,  200,  100,  100,  300,  300],
    'PLAYER_ID':  [1,    2,    3,    4,    1,    2,    5,    6],
    'FGM':        [10,   4,    2,    8,    10,   11,   8,    7],
    'FG3M':       [12,   4,    6,    2,    4,    5,    10,   6],
    'FTM':        [12,   7,    5,    7,    10,   4,    9,    3]
})

# Create and print the Boxscore object
box = Boxscore(data=df)
print(box)
print("-" * 40) # Separator
